sdk/core/azure-core/src/main/java/com/azure/core/util/Option.java

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+/**
+ * The Option type, every Option is either has a value present including null,
+ * or has no value present.
+ *
+ * @param <T> The value type.
+ */
+public final class Option<T> {
+    private static final Option<?> UNSET = new Option<>();
+    private final boolean isSet;
+    private final T value;
+
+    /**
+     * Returns an {@link Option} instance with no value.
+     *
+     * @param <T> The value type.
+     * @return An Option type with no value present.
+     */
+    public static <T> Option<T> unset() {
+        @SuppressWarnings("unchecked")
+        Option<T> none = (Option<T>) UNSET;
+        return none;
+    }
+
+    /**
+     * Returns an {@link Option} with an empty value.
+     *
+     * @param <T> The value type.
+     * @return an {@link Option} with the empty value present.
+     */
+    public static <T> Option<T> empty() {
+        return new Option<>(null);
+    }
+
+    /**
+     * Returns an {@link Option} with the specified non-null value.
+     *
+     * @param <T> The value type.
+     * @param value the value, which must be non-null.
+     * @return an {@link Option} with the non-null value present.
+     * @throws NullPointerException if value is null.
+     */
+    public static <T> Option<T> of(T value) {
+        Objects.requireNonNull(value);
+        return new Option<>(value);
+    }
+
+    /**
+     * Returns an {@link Option} with the specified nullable value.
+     *
+     * @param <T> The value type.
+     * @param value the null or non-null value.
+     * @return an {@link Option} with the value present.
+     */
+    public static <T> Option<T> ofNullable(T value) {
+        return new Option<>(value);
+    }
+
+    /**
+     * Return {@code true} if there is a value present, otherwise {@code false}.
+     *
+     * @return {@code true} if there is a value present, otherwise {@code false}
+     */
+    public boolean isSet() {
+        return this.isSet;
+    }
```

Comment:
If we use the JS terminology, we can change this method name to `isInitialized()`. We should also consider adding `isPresent()` to return `true` if `isInitialized() && value != null`.

Reply:
sounds good, so `Option.unset()` will be `Option.uninitialized()`. `Option.isSet()` become `Option.isInitialized()`

Reply:
I've question-related to `isPresent()`, given "of(null)" represent "presence" of explicit-null, would it be confusing having this method returning `false` when Option is in the explicit-null state?

Reply:
I think that's what `isPresent()` in Java's Optional does. It returns true only if the value is non-null. For everything else, it returns false. So, in our case, isPresent will return the following:

uninitialized - `false`
null value - `false`
non-null value - `true`

Reply:
Also, if you think there's room for confusion with this API and is something that we don't need to add right away. We can add this later too. It just means that the user will have to do this check themselves: `if(patchValue.isInitialized() && patchValue.getValue() != null)`. So, `isPresent()` is just syntactic sugar that can be introduced later, if required.

Reply:
thanks! yes, it's clear the purpose of "isPresent()" w.r.t `Optional`, given it's a two-state type. It was a bit confusing for me the behaviour of  "isPresent()" in a tri-state type. 


Like you said, maybe we want to hold on to it and introduce it later once we see a need for it. At least now, I see the "Option" type is mainly used in the input model than output model, its mostly inspected in the azure-core side while serializing.

Reply:
Renamed, `Option.unset()` -> `Option.uninitialized()`. `Option.isSet()` -> `Option.isInitialized()`.